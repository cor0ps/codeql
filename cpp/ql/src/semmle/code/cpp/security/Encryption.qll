// Common predicates relating to encryption in C and C++
import cpp

/** A blacklist of algorithms that are known to be insecure */
string algorithmBlacklist() {
  result = "DES" or
  result = "RC2" or
  result = "RC4" or
  result = "RC5" or
  result = "ARCFOUR" // a variant of RC4
}

// these are only bad if they're being used for encryption, and it's
// hard to know when that's happening
string hashAlgorithmBlacklist() {
  result = "SHA1" or
  result = "MD5"
}

/** A regex for matching strings that look like they contain a blacklisted algorithm */
string algorithmBlacklistRegex() {
  result =
    // algorithms usually appear in names surrounded by characters that are not
    // alphabetical characters in the same case. This handles the upper and lower
    // case cases
    "(^|.*[^A-Z])(" + strictconcat(algorithmBlacklist(), "|") + ")([^A-Z].*|$)" + "|" +
      // for lowercase, we want to be careful to avoid being confused by camelCase
      // hence we require two preceding uppercase letters to be sure of a case switch,
      // or a preceding non-alphabetic character
      "(^|.*[A-Z]{2}|.*[^a-zA-Z])(" + strictconcat(algorithmBlacklist().toLowerCase(), "|") +
      ")([^a-z].*|$)"
}

/** A whitelist of algorithms that are known to be secure */
string algorithmWhitelist() {
  result = "RSA" or
  result = "SHA256" or
  result = "CCM" or
  result = "GCM" or
  result = "AES" or
  result = "Blowfish" or
  result = "ECIES"
}

/** A regex for matching strings that look like they contain a whitelisted algorithm */
string algorithmWhitelistRegex() {
  // The implementation of this is a duplicate of algorithmBlacklistRegex, as it isn't
  // possible to have string -> string functions at the moment
  // algorithms usually appear in names surrounded by characters that are not
  // alphabetical characters in the same case. This handles the upper and lower
  // case cases
  result = "(^|.*[^A-Z])" + algorithmWhitelist() + "([^A-Z].*|$)"
  or
  // for lowercase, we want to be careful to avoid being confused by camelCase
  // hence we require two preceding uppercase letters to be sure of a case switch,
  // or a preceding non-alphabetic character
  result = "(^|.*[A-Z]{2}|.*[^a-zA-Z])" + algorithmWhitelist().toLowerCase() + "([^a-z].*|$)"
}
