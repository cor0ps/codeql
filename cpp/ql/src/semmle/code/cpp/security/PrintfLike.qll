import semmle.code.cpp.commons.Printf
import external.ExternalArtifact

predicate printfLikeFunction(Function func, int formatArg) {
  formatArg = func.(FormattingFunction).getFormatParameterIndex() and
  not func instanceof UserDefinedFormattingFunction
  or
  primitiveVariadicFormatter(func, formatArg)
  or
  exists(ExternalData data |
    // TODO Do this \ to / conversion in the toolchain?
    data.getDataPath().replaceAll("\\", "/") = "cert/formatingFunction.csv" and
    func.getName() = data.getField(0) and
    formatArg = data.getFieldAsInt(1)
  )
}
