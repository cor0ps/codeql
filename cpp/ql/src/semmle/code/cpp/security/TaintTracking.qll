/*
 * Support for tracking tainted data through the program.
 */

import semmle.code.cpp.ir.dataflow.DefaultTaintTracking
