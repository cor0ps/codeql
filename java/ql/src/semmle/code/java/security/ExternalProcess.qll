/* Definitions related to external processes. */
import semmle.code.java.Member
import semmle.code.java.JDK
import semmle.code.java.frameworks.apache.Exec

/**
 * An expression used as an argument to a call that executes an external command. For calls to
 * varargs method calls, this only includes the first argument, which will be the command
 * to be executed.
 */
class ArgumentToExec extends Expr {
  ArgumentToExec() {
    exists(MethodAccess execCall, Method method |
      execCall.getArgument(0) = this and
      method = execCall.getMethod() and
      (
        method instanceof MethodRuntimeExec or
        method instanceof MethodProcessBuilderCommand or
        method instanceof MethodCommandLineParse or
        method instanceof MethodCommandLineAddArguments
      )
    )
    or
    exists(ConstructorCall expr, Constructor cons |
      expr.getConstructor() = cons and
      cons.getDeclaringType().hasQualifiedName("java.lang", "ProcessBuilder") and
      expr.getArgument(0) = this
    )
  }
}

/**
 * An `ArgumentToExec` of type `String`.
 */
class StringArgumentToExec extends ArgumentToExec {
  StringArgumentToExec() { this.getType() instanceof TypeString }
}
