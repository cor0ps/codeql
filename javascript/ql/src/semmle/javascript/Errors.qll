/** Provides classes for working with syntax errors. */

import javascript

/** An error encountered during extraction. */
abstract class Error extends Locatable {
  override Location getLocation() { hasLocation(this, result) }

  /** Gets the message associated with this error. */
  abstract string getMessage();

  override string toString() { result = getMessage() }
}

/** A JavaScript parse error encountered during extraction. */
class JSParseError extends @js_parse_error, Error {
  /** Gets the toplevel element this error occurs in. */
  TopLevel getTopLevel() { jsParseErrors(this, result, _, _) }

  override string getMessage() { jsParseErrors(this, _, result, _) }

  /** Gets the source text of the line this error occurs on. */
  string getLine() { jsParseErrors(this, _, _, result) }
}
