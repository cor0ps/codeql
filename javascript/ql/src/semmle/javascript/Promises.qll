/**
 * Provides classes for modelling promises and their data-flow.
 */

import javascript

/**
 * A definition of a `Promise` object.
 */
abstract class PromiseDefinition extends DataFlow::SourceNode {
  /** Gets the executor function of this promise object. */
  abstract DataFlow::FunctionNode getExecutor();

  /** Gets the `resolve` parameter of the executor function. */
  DataFlow::ParameterNode getResolveParameter() { result = getExecutor().getParameter(0) }

  /** Gets the `reject` parameter of the executor function. */
  DataFlow::ParameterNode getRejectParameter() { result = getExecutor().getParameter(1) }

  /** Gets the `i`th callback handler installed by method `m`. */
  private DataFlow::FunctionNode getAHandler(string m, int i) {
    result = getAMethodCall(m).getCallback(i)
  }

  /**
   * Gets a function that handles promise resolution, including both
   * `then` handlers and `finally` handlers.
   */
  DataFlow::FunctionNode getAResolveHandler() {
    result = getAHandler("then", 0) or
    result = getAFinallyHandler()
  }

  /**
   * Gets a function that handles promise rejection, including
   * `then` handlers, `catch` handlers and `finally` handlers.
   */
  DataFlow::FunctionNode getARejectHandler() {
    result = getAHandler("then", 1) or
    result = getACatchHandler() or
    result = getAFinallyHandler()
  }

  /**
   * Gets a `catch` handler of this promise.
   */
  DataFlow::FunctionNode getACatchHandler() { result = getAHandler("catch", 0) }

  /**
   * Gets a `finally` handler of this promise.
   */
  DataFlow::FunctionNode getAFinallyHandler() { result = getAHandler("finally", 0) }
}

/** Holds if the `i`th callback handler is installed by method `m`. */
private predicate hasHandler(DataFlow::InvokeNode promise, string m, int i) {
  exists(promise.getAMethodCall(m).getCallback(i))
}

/**
 * A call that looks like a Promise.
 *
 * For example, this could be the call `promise(f).then(function(v){...})`
 */
class PromiseCandidate extends DataFlow::InvokeNode {
  PromiseCandidate() {
    hasHandler(this, "then", [0 .. 1]) or
    hasHandler(this, "catch", 0) or
    hasHandler(this, "finally", 0)
  }
}

/**
 * A promise object created by the standard ECMAScript 2015 `Promise` constructor.
 */
private class ES2015PromiseDefinition extends PromiseDefinition, DataFlow::NewNode {
  ES2015PromiseDefinition() { this = DataFlow::globalVarRef("Promise").getAnInstantiation() }

  override DataFlow::FunctionNode getExecutor() { result = getCallback(0) }
}

/**
 * A promise that is created and resolved with one or more value.
 */
abstract class PromiseCreationCall extends DataFlow::CallNode {
  /**
   * Gets the value this promise is resolved with.
   */
  abstract DataFlow::Node getValue();
}

/**
 * A promise that is created using a `.resolve()` call.
 */
abstract class ResolvedPromiseDefinition extends PromiseCreationCall { }

/**
 * A resolved promise created by the standard ECMAScript 2015 `Promise.resolve` function.
 */
class ResolvedES2015PromiseDefinition extends ResolvedPromiseDefinition {
  ResolvedES2015PromiseDefinition() {
    this = DataFlow::globalVarRef("Promise").getAMemberCall("resolve")
  }

  override DataFlow::Node getValue() { result = getArgument(0) }
}

/**
 * An aggregated promise produced either by `Promise.all` or `Promise.race`.
 */
class AggregateES2015PromiseDefinition extends PromiseCreationCall {
  AggregateES2015PromiseDefinition() {
    exists(string m | m = "all" or m = "race" |
      this = DataFlow::globalVarRef("Promise").getAMemberCall(m)
    )
  }

  override DataFlow::Node getValue() {
    result = getArgument(0).getALocalSource().(DataFlow::ArrayCreationNode).getAnElement()
  }
}

/**
 * This module defines how data-flow propagates into and out of a Promise.
 * The data-flow is based on pseudo-properties rather than tainting the Promise object (which is what `PromiseTaintStep` does).
 */
private module PromiseFlow {
  /**
   * Gets the pseudo-field used to describe resolved values in a promise.
   */
  string resolveField() { result = "$PromiseResolveField$" }

  /**
   * Gets the pseudo-field used to describe rejected values in a promise.
   */
  string rejectField() { result = "$PromiseRejectField$" }

  /**
   * A flow step describing a promise definition.
   *
   * The resolved/rejected value is written to a pseudo-field on the promise.
   */
  class PromiseDefitionStep extends DataFlow::AdditionalFlowStep {
    PromiseDefinition promise;

    PromiseDefitionStep() { this = promise }

    override predicate storeStep(DataFlow::Node pred, DataFlow::Node succ, string prop) {
      prop = resolveField() and
      pred = promise.getResolveParameter().getACall().getArgument(0) and
      succ = this
      or
      prop = rejectField() and
      (
        pred = promise.getRejectParameter().getACall().getArgument(0) or
        pred = promise.getExecutor().getExceptionalReturn()
      ) and
      succ = this
    }

    override predicate loadStoreStep(DataFlow::Node pred, DataFlow::Node succ, string prop) {
      // Copy the value of a resolved promise to the value of this promise.
      prop = resolveField() and
      pred = promise.getResolveParameter().getACall().getArgument(0) and
      succ = this
    }
  }

  /**
   * A flow step describing the a Promise.resolve (and similar) call.
   */
  class CreationStep extends DataFlow::AdditionalFlowStep {
    PromiseCreationCall promise;

    CreationStep() { this = promise }

    override predicate storeStep(DataFlow::Node pred, DataFlow::Node succ, string prop) {
      prop = resolveField() and
      pred = promise.getValue() and
      succ = this
    }

    override predicate loadStoreStep(DataFlow::Node pred, DataFlow::Node succ, string prop) {
      // Copy the value of a resolved promise to the value of this promise.
      prop = resolveField() and
      pred = promise.getValue() and
      succ = this
    }
  }

  /**
   * A load step loading the pseudo-field describing that the promise is rejected.
   * The rejected value is thrown as a exception.
   */
  class AwaitStep extends DataFlow::AdditionalFlowStep {
    DataFlow::Node operand;
    AwaitExpr await;

    AwaitStep() {
      this.getEnclosingExpr() = await and
      operand.getEnclosingExpr() = await.getOperand()
    }

    override predicate loadStep(DataFlow::Node pred, DataFlow::Node succ, string prop) {
      prop = resolveField() and
      succ = this and
      pred = operand
      or
      prop = rejectField() and
      succ = await.getExceptionTarget() and
      pred = operand
    }
  }

  /**
   * A flow step describing the data-flow related to the `.then` method of a promise.
   */
  class ThenStep extends DataFlow::AdditionalFlowStep, DataFlow::MethodCallNode {
    ThenStep() { this.getMethodName() = "then" }

    override predicate loadStep(DataFlow::Node pred, DataFlow::Node succ, string prop) {
      prop = resolveField() and
      pred = getReceiver() and
      succ = getCallback(0).getParameter(0)
      or
      prop = rejectField() and
      pred = getReceiver() and
      succ = getCallback(1).getParameter(0)
    }

    override predicate loadStoreStep(DataFlow::Node pred, DataFlow::Node succ, string prop) {
      not exists(this.getArgument(1)) and
      prop = rejectField() and
      pred = getReceiver() and
      succ = this
      or
      // read the value of a resolved/rejected promise that is returned
      (prop = rejectField() or prop = resolveField()) and
      pred = getCallback([0 .. 1]).getAReturn() and
      succ = this
    }

    override predicate storeStep(DataFlow::Node pred, DataFlow::Node succ, string prop) {
      prop = resolveField() and
      pred = getCallback([0 .. 1]).getAReturn() and
      succ = this
      or
      prop = rejectField() and
      pred = getCallback([0 .. 1]).getExceptionalReturn() and
      succ = this
    }
  }

  /**
   * A flow step describing the data-flow related to the `.catch` method of a promise.
   */
  class CatchStep extends DataFlow::AdditionalFlowStep, DataFlow::MethodCallNode {
    CatchStep() { this.getMethodName() = "catch" }

    override predicate loadStep(DataFlow::Node pred, DataFlow::Node succ, string prop) {
      prop = rejectField() and
      pred = getReceiver() and
      succ = getCallback(0).getParameter(0)
    }

    override predicate loadStoreStep(DataFlow::Node pred, DataFlow::Node succ, string prop) {
      prop = resolveField() and
      pred = getReceiver().getALocalSource() and
      succ = this
      or
      // read the value of a resolved/rejected promise that is returned
      (prop = rejectField() or prop = resolveField()) and
      pred = getCallback(0).getAReturn() and
      succ = this
    }

    override predicate storeStep(DataFlow::Node pred, DataFlow::Node succ, string prop) {
      prop = rejectField() and
      pred = getCallback(0).getExceptionalReturn() and
      succ = this
      or
      prop = resolveField() and
      pred = getCallback(0).getAReturn() and
      succ = this
    }
  }

  /**
   * A flow step describing the data-flow related to the `.finally` method of a promise.
   */
  class FinallyStep extends DataFlow::AdditionalFlowStep, DataFlow::MethodCallNode {
    FinallyStep() { this.getMethodName() = "finally" }

    override predicate loadStoreStep(DataFlow::Node pred, DataFlow::Node succ, string prop) {
      (prop = resolveField() or prop = rejectField()) and
      pred = getReceiver() and
      succ = this
      or
      // read the value of a rejected promise that is returned
      prop = rejectField() and
      pred = getCallback(0).getAReturn() and
      succ = this
    }

    override predicate storeStep(DataFlow::Node pred, DataFlow::Node succ, string prop) {
      prop = rejectField() and
      pred = getCallback(0).getExceptionalReturn() and
      succ = this
    }
  }
}

/**
 * Holds if taint propagates from `pred` to `succ` through promises.
 */
predicate promiseTaintStep(DataFlow::Node pred, DataFlow::Node succ) {
  // from `x` to `new Promise((res, rej) => res(x))`
  pred = succ.(PromiseDefinition).getResolveParameter().getACall().getArgument(0)
  or
  // from `x` to `Promise.resolve(x)`
  pred = succ.(PromiseCreationCall).getValue()
  or
  exists(DataFlow::MethodCallNode thn | thn.getMethodName() = "then" |
    // from `p` to `x` in `p.then(x => ...)`
    pred = thn.getReceiver() and
    succ = thn.getCallback(0).getParameter(0)
    or
    // from `v` to `p.then(x => return v)`
    pred = thn.getCallback([0 .. 1]).getAReturn() and
    succ = thn
  )
  or
  exists(DataFlow::MethodCallNode catch | catch.getMethodName() = "catch" |
    // from `p` to `p.catch(..)`
    pred = catch.getReceiver() and
    succ = catch
    or
    // from `v` to `p.catch(x => return v)`
    pred = catch.getCallback(0).getAReturn() and
    succ = catch
  )
  or
  // from `p` to `p.finally(..)`
  exists(DataFlow::MethodCallNode finally | finally.getMethodName() = "finally" |
    pred = finally.getReceiver() and
    succ = finally
  )
  or
  // from `x` to `await x`
  exists(AwaitExpr await |
    pred.getEnclosingExpr() = await.getOperand() and
    succ.getEnclosingExpr() = await
  )
}

/**
 * An additional taint step that involves promises.
 */
private class PromiseTaintStep extends TaintTracking::AdditionalTaintStep {
  DataFlow::Node source;

  PromiseTaintStep() { promiseTaintStep(source, this) }

  override predicate step(DataFlow::Node pred, DataFlow::Node succ) {
    pred = source and succ = this
  }
}

/**
 * Provides classes for working with the `bluebird` library (http://bluebirdjs.com).
 */
module Bluebird {
  private DataFlow::SourceNode bluebird() {
    result = DataFlow::globalVarRef("Promise") or // same as ES2015PromiseDefinition!
    result = DataFlow::moduleImport("bluebird")
  }

  /**
   * A promise object created by the bluebird `Promise` constructor.
   */
  private class BluebirdPromiseDefinition extends PromiseDefinition, DataFlow::NewNode {
    BluebirdPromiseDefinition() { this = bluebird().getAnInstantiation() }

    override DataFlow::FunctionNode getExecutor() { result = getCallback(0) }
  }

  /**
   * A resolved promise created by the bluebird `Promise.resolve` function.
   */
  class ResolvedBluebidPromiseDefinition extends ResolvedPromiseDefinition {
    ResolvedBluebidPromiseDefinition() { this = bluebird().getAMemberCall("resolve") }

    override DataFlow::Node getValue() { result = getArgument(0) }
  }

  /**
   * An aggregated promise produced either by `Promise.all`, `Promise.race` or `Promise.map`.
   */
  class AggregateBluebirdPromiseDefinition extends PromiseCreationCall {
    AggregateBluebirdPromiseDefinition() {
      exists(string m | m = "all" or m = "race" or m = "map" | this = bluebird().getAMemberCall(m))
    }

    override DataFlow::Node getValue() {
      result = getArgument(0).getALocalSource().(DataFlow::ArrayCreationNode).getAnElement()
    }
  }
}

/**
 * Provides classes for working with the `q` library (https://github.com/kriskowal/q).
 */
module Q {
  /**
   * A promise object created by the q `Promise` constructor.
   */
  private class QPromiseDefinition extends PromiseDefinition, DataFlow::CallNode {
    QPromiseDefinition() { this = DataFlow::moduleMember("q", "Promise").getACall() }

    override DataFlow::FunctionNode getExecutor() { result = getCallback(0) }
  }
}

private module ClosurePromise {
  /**
   * A promise created by a call `new goog.Promise(executor)`.
   */
  private class ClosurePromiseDefinition extends PromiseDefinition, DataFlow::NewNode {
    ClosurePromiseDefinition() { this = Closure::moduleImport("goog.Promise").getACall() }

    override DataFlow::FunctionNode getExecutor() { result = getCallback(0) }
  }

  /**
   * A promise created by a call `goog.Promise.resolve(value)`.
   */
  private class ResolvedClosurePromiseDefinition extends ResolvedPromiseDefinition {
    ResolvedClosurePromiseDefinition() {
      this = Closure::moduleImport("goog.Promise.resolve").getACall()
    }

    override DataFlow::Node getValue() { result = getArgument(0) }
  }

  /**
   * Taint steps through closure promise methods.
   */
  private class ClosurePromiseTaintStep extends TaintTracking::AdditionalTaintStep {
    DataFlow::Node pred;

    ClosurePromiseTaintStep() {
      // static methods in goog.Promise
      exists(DataFlow::CallNode call, string name |
        call = Closure::moduleImport("goog.Promise." + name).getACall() and
        this = call and
        pred = call.getAnArgument()
      |
        name = "all" or
        name = "allSettled" or
        name = "firstFulfilled" or
        name = "race"
      )
      or
      // promise created through goog.promise.withResolver()
      exists(DataFlow::CallNode resolver |
        resolver = Closure::moduleImport("goog.Promise.withResolver").getACall() and
        this = resolver.getAPropertyRead("promise") and
        pred = resolver.getAMethodCall("resolve").getArgument(0)
      )
    }

    override predicate step(DataFlow::Node src, DataFlow::Node dst) { src = pred and dst = this }
  }
}
