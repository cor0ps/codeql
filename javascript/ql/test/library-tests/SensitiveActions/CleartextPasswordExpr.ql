import javascript
import semmle.javascript.security.SensitiveActions

from CleartextPasswordExpr e
select e
