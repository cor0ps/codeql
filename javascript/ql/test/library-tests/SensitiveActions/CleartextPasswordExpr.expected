| tst.js:1:1:1:8 | password |
| tst.js:2:1:2:8 | PassWord |
| tst.js:3:1:3:21 | myPassw ... eartext |
| tst.js:4:1:4:10 | x.password |
| tst.js:5:1:5:11 | getPassword |
| tst.js:5:1:5:13 | getPassword() |
| tst.js:6:1:6:13 | x.getPassword |
| tst.js:6:1:6:15 | x.getPassword() |
| tst.js:7:1:7:15 | get("password") |
| tst.js:8:1:8:17 | x.get("password") |
